Make Apache HTTP client more configurable and improve default performances

aseure opened the pull request:
Fix #678

aseure committed:
feat: deprecate ConfigBase Builder.setUseSystemProxy() method

aseure committed:
feat: expose new constructor for the ApacheHttpRequester

To ease the manual configuration of the internal Apache async HTTP client,
this commit adds a new constructor to the `ApacheHttpRequester` which
accepts an `HttpAsyncClientBuilder` instance. This builder will be used
as-is to generate the configuration used by the underlying HTTP client.

Fix #678

aseure committed:
doc(README): provide a how-to for configuring the Apache HTTP client manually

Ant-hem commented on algoliasearch-core/src/main/java/com/algolia/search/ConfigBase.java:
```diff
@@ -82,6 +82,7 @@ public Builder(
     public abstract T getThis();
 
     /** Makes the underlying HTTP Client to use JVM/System settings for the proxy */
+    @Deprecated
```
I would add a note in the deprecation attribute telling to use the `builder` instead.

aseure replied:
`@Deprecated` annotation doesn't support optional attributes before Java 9.

I'll put a comment on the same line instead!

Ant-hem replied:
I've done this like this before:

https://github.com/algolia/algoliasearch-client-java-2/blob/6289d61dd96bad7dd3c298dd8b4c24e1ab1bc0ce/algoliasearch-core/src/main/java/com/algolia/search/SearchClientPersonalization.java#L19-L27

aseure replied:
OH, even better, I didn't remember of the javadoc way of doing it. Thanks! Changing it right now.

aseure committed:
feat(apache): improve performances by changing maxConnPerRoute from 2 to 100

After some investigations in #678, we discovered that our default value
of `maxConnPerRoute` was set to 2. This led to poor performances
compared to our previous v1 major version. After some benchmarking, we
noticed that this default value, while being conservative, is probably
to low for most use-cases.

Note that we can safely change this value since it can now be overriden
by passing a custom `HttpAsyncClientBuilder` during the `ApacheHttpRequester`
instanciation (see our README for more information).

Ant-hem commented on algoliasearch-apache/src/main/java/com/algolia/search/ApacheHttpRequester.java:
```diff
@@ -31,14 +32,13 @@
 
-  ApacheHttpRequester(@Nonnull ConfigBase config) {
-
+  public ApacheHttpRequester(@Nonnull ConfigBase config) {
```
nit: We could call `ApacheHttpRequester(@Nonnull ConfigBase config, @Nonnull HttpAsyncClientBuilder builder)` with the builder created in place in `ApacheHttpRequester(@Nonnull ConfigBase config)` to avoid duplicated code in the ctors.

aseure replied:
I wanted to do it at first but then abandoned because I needed to pass the big ternary in a `this( ... );` statement which must be the first immediate statement in the constructor to be able to chain them. But after retrying it, it seems kinda clean anyway so I've did it in the end.

Ant-hem reviewed (approved):
Awesome, thanks! Let's go.

aseure committed:
test: fix flaky ReplacingTest

aseure committed:
chore: replace TravisCI with CircleCI

Ant-hem reviewed (approved):
Let's go! Thanks for migrating to Circle.
